File: src/org/grumblesmurf/df/careers/Dwarf.java
package org.grumblesmurf.df.careers;

import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import javax.swing.text.BadLocationException;
import javax.swing.text.html.HTMLDocument;
import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class Dwarf implements Comparable<Dwarf>
{
    final String name;
    private final String nickName;
    final String sex;
    private final Map<String, Integer> attributes;
    private final List<String> traits;
    private final String displayName;

    public Dwarf(String name, String nickName, String sex, Map<String, Integer> attributes, List<String> traits) {
        this.name = name;
        this.nickName = nickName;
        this.sex = sex;
        this.attributes = attributes;
        this.traits = traits;

        if (nickName.isEmpty()) {
            this.displayName = name;
        } else {
            this.displayName = String.format("%s (%s)", name, nickName);
        }
    }

    public static Dwarf from(Element creature) {
        String name = textContentOfFirstChildNamed("Name", creature);
        String nickName = textContentOfFirstChildNamed("Nickname", creature);
        String sex = textContentOfFirstChildNamed("Sex", creature);

        Element attributesElement = firstChildNamed("Attributes", creature);
        NodeList childNodes = attributesElement.getChildNodes();
        Map<String, Integer> attributes = new LinkedHashMap<String, Integer>();
        for (int i = 0; i < childNodes.getLength(); i++) {
            Node item = childNodes.item(i);
            if (item instanceof Element) {
                attributes.put(item.getNodeName(), Integer.valueOf(item.getTextContent()));
            }
        }

        NodeList traitsElements = creature.getElementsByTagName("Trait");
        List<String> traits = new LinkedList<String>();
        for (int i = 0; i < traitsElements.getLength(); i++) {
            Node trait = traitsElements.item(i);
            traits.add(trait.getTextContent());
        }

        return new Dwarf(name, nickName, sex, attributes, traits);
    }

    private static Element firstChildNamed(String name, Element parent) {
        return (Element) parent.getElementsByTagName(name).item(0);
    }

    private static String textContentOfFirstChildNamed(String name, Element parent) {
        return firstChildNamed(name, parent).getTextContent();
    }

    @Override
    public int compareTo(Dwarf o) {
        return displayName.compareTo(o.displayName);
    }

    @Override
    public String toString() {
        return displayName;
    }

    public void renderTo(javax.swing.text.Element body)
        throws IOException, BadLocationException {
        HTMLDocument d = (HTMLDocument) body.getDocument();
        d.insertBeforeEnd(body, nameHeader());
        d.insertBeforeEnd(body, mainLayout());
        d.insertBeforeEnd(d.getElement("left"), attributesTable());
        d.insertBeforeEnd(d.getElement("right"), traitsList());
    }

    private String mainLayout() {
        return "<table valign='top'><tr><td id='left'></td><td id='right'></td></tr></table>";
    }

    private String nameHeader() {
        return String.format("<h1>%s</h1>", displayName);
    }

    private String attributesTable() {
        StringBuilder sb = new StringBuilder("<table>");
        for (Map.Entry<String, Integer> entry : attributes.entrySet()) {
            sb.append("<tr>")
                .append("<th>").append(entry.getKey()).append("</th>")
                .append("<td>").append(entry.getValue()).append("</td>")
                .append("</tr>");
        }

        return sb.append("</table>").toString();
    }

    private String traitsList() {
        StringBuilder sb = new StringBuilder("<ul>");
        for (String trait : traits) {
            sb.append("<li>").append(trait).append("</li>");
        }
        return sb.append("</ul>").toString();
    }
}

File: src/org/grumblesmurf/df/careers/Dwarf.scala
package org.grumblesmurf.df.careers

import javax.swing.text.html.HTMLDocument
import xml.{Elem, Node}
import scala.collection.SortedMap

object Dwarf {
  def from(node: Node): Dwarf = {
    var name = (node \ "Name").text
    val bits = name.split(" ")
    if (bits.length > 1 && (bits(0) == bits(1))) {
      name = name.substring(bits(0).length + 1)
    }
    val nickName = (node \ "Nickname").text
    val sex = (node \ "Sex").text
    val attributes = SortedMap.empty[Attribute,Int] ++ (node \ "Attributes").head.nonEmptyChildren.collect {
      case a: Elem =>
        Attribute.valueOf(a.label) -> a.text.toInt
    }.toMap

    val traits = SortedMap.empty[String,Trait] ++ (node \\ "Trait").map { t =>
      (t \ "@name").text -> Trait((t \ "@value").text.toInt, t.text)
    }.toMap

    new Dwarf(name, nickName, sex, attributes, traits)
  }
}

case class Trait(value: Int, description: String)

class Dwarf(val name: String, val nickName: String, val sex: String, val attributes: Map[Attribute, Int], val traits: Map[String, Trait]) extends Comparable[Dwarf] {
  val displayName: String = if (nickName.isEmpty) name else "%s (%s)" format (name, nickName)

  def compareTo(o: Dwarf): Int = displayName.compareTo(o.displayName)

  override def toString: String = displayName

  def renderTo(body: javax.swing.text.Element) {
    val d = body.getDocument.asInstanceOf[HTMLDocument]
    d.insertBeforeEnd(body, nameHeader)
    d.insertBeforeEnd(body, mainLayout)
    d.insertBeforeEnd(d.getElement("left"), attributesTable)
    d.insertBeforeEnd(d.getElement("right"), traitsList)
    d.insertBeforeEnd(d.getElement("right"), positionList)
  }

  private def positionList = {
    val evaluations = SortedMap.empty[Evaluation,Array[Position]] ++ Position.values().map(p => p.evaluate(this) -> p).groupBy(_._1).map {
      case (e, ps) => e -> ps.map(_._2)
    }
    evaluations.map {
      case (evaluation, positions) =>
        <div>
          <h2>{evaluation}</h2>
          <ul>
            {positions.map {p => <li>{p}</li> }}
          </ul>
        </div>
    }.mkString("\n")
  }

  private def mainLayout = <table valign='top'>
    <tr>
      <td id='left'></td> <td id='right'></td>
    </tr>
  </table>.toString()

  private def nameHeader = {
    <h1>
      {displayName}
    </h1>.toString()
  }

  private def attributesTable = {
    <table>
      {attributes.map{
      case (attr, value) => <tr><th>{attr}</th><td>{value}</td></tr>
    }}
    </table>.toString()
  }

  private def traitsList = {
    <div>
      <ul>
        {traits.values.filter(_.description.nonEmpty).map(t => <li>{t.description}</li>)}
      </ul>
      <hr/>
    </div>.toString()
  }

  def attribute(key: String) = attributes(Attribute.valueOf(key))

  def getTrait(key: String) = traits(key)

  override def equals(o: Any): Boolean = o.isInstanceOf[Dwarf] && compareTo(o.asInstanceOf[Dwarf]) == 0

  override def hashCode: Int = {
    if (displayName != null) displayName.hashCode else 0
  }
}
